### PR TITLE
feat: `#[derive(LightHasher)]` macro and `AsByteVec` trait

### PR DESCRIPTION
**Pull request by vadorovsky**
- `AsByteVec`, providing `as_byte_vec()` method, guarantees consistent way of serializing types to 2D byte vectors before hashing. We provide default implementations for primitives. More complex types need to implement `IntoBytes` themselves.
  - By using 2D vectors, we make it possible to represent structs with multiple fields. This way, we can handle struct fields (nested structs) the same way as primitive fields and deal with all types by using one trait.
  - The reason behing using vectors instead of slices is that in some cases, we cannot just cast the type without defining custom bytes. Vectors, although they might introduce copies, make sure we always own the bytes we are creating.
- `#[derive(LightHasher)]` implements `AsByteVec` and `DataHasher` traits. The `DataHasher` implementation takes bytes returned by `AsByteVec` as an input.

### CODE REVIEW

File: merkle-tree/hasher/src/bytes.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        assert_eq!(u16_max_bytes, &[u16::MAX.to_ne_bytes()]);
+
+        let i32_min: &dyn AsByteVec = &i32::MIN;
+        let i32_min_bytes = i32_min.as_byte_vec();
+        assert_eq!(i32_min_bytes, &[&[0, 0, 0, 128]]);
+        assert_eq!(i32_min_bytes, &[i32::MIN.to_ne_bytes()]);
+        let i32_max: &dyn AsByteVec = &i32::MAX;
+        let i32_max_bytes = i32_max.as_byte_vec();
+        assert_eq!(i32_max_bytes, &[&[255, 255, 255, 127]]);
+        assert_eq!(i32_max_bytes, &[i32::MAX.to_ne_bytes()]);
```
Comment (SwenSchaeferjohann) [inline merkle-tree/hasher/src/bytes.rs:143]: 1) aren't these all redundant assertions? 
2) provided that `as_byte_vec` returns a vec, shouldn't the assertion be against `&[vec![255, 255, 255, 127]]` ?
@vadorovsky

File: merkle-tree/hasher/src/bytes.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+#[cfg(feature = "solana")]
+impl AsByteVec for solana_program::pubkey::Pubkey {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.to_bytes().to_vec()]
+    }
+}
+
+#[cfg(test)]
```
Comment (SwenSchaeferjohann) [inline merkle-tree/hasher/src/bytes.rs:94]: would it make sense to add test cases for additional edge cases such as:
* string / empty string
* option<T> with none, some
* empty array

File: macros/light/src/lib.rs [hunk 3]
```diff
@@ -433,3 +434,11 @@ pub fn light_discriminator(input: TokenStream) -> TokenStream {
         .unwrap_or_else(|err| err.to_compile_error())
         .into()
 }
+
```
Comment (SwenSchaeferjohann) [inline macros/light/src/lib.rs:437]: Can we add an inline doc similar to the documentation for the `light_traits` and `light_accounts` macros?

The goal is: developers read only this and understand how to use it successfully in their applications.

Comment (vadorovsky) [reply macros/light/src/lib.rs:437]: Yes, will do

File: merkle-tree/hasher/src/bytes.rs (hunk 1 above)
Comment (vadorovsky) [reply merkle-tree/hasher/src/bytes.rs:143]: 1. Well, in a way they are, but I just wanted to assure myself that `as_byte_vec` on primitives returns the same values as the std functions, but also explicitly check the exact values.
2. In Rust, you can compare vectors with slices, so these assertions work. :)

File: merkle-tree/hasher/src/bytes.rs (hunk 2 above)
Comment (vadorovsky) [reply merkle-tree/hasher/src/bytes.rs:94]: Done

File: macros/light/src/lib.rs (hunk 3 above)
Comment (vadorovsky) [reply macros/light/src/lib.rs:165]: Done

**Review by SwenSchaeferjohann — APPROVED** [review]
